### ISSUE TITLE
Support syncing metadata forcibly when accessing the file for the first time for the client

### ISSUE THREAD

**Issue by secfree** [Alluxio/alluxio#16924]
**Is your feature request related to a problem? Please describe.**

Some of our alluxio-fuse users feedback that their training programs have bad performance while

1. The program launches a lot of threads reading the same path (eg. listStatus the same directory) AND
2. Set `alluxio.user.file.metadata.sync.interval=0` to get the latest data

This is expected as

1. For each request, Alluxio needs to reach UFS to check if anything changed AND
2. AlluxioMaster needs to place a WRITE lock on the corresponding node (sync metadata may trigger updating) and then each request needs to wait

After communicating with the users, we got that

1. When starting their programs, they knew that the dependent data has updated/finished in UFS
2. But they were not sure whether the data in Alluxio has synced, so they set `alluxio.user.file.metadata.sync.interval=0` to initialize the alluxio-fuse container

For this case, setting `alluxio.user.file.metadata.sync.interval=0` is not needed and is too expensive. What it needs is just syncing metadata once before launching the program. But as they are alluxio-fuse users and Alluxio is transparent to some of them, it's not easy/convenient for them to "sync recursively once".

**Describe the solution you'd like**

One solution is to support syncing metadata forcibly when accessing the file for the first time for the client
1. The client uses BloomFilter to record paths that it has accessed
2. If a path is already accessed, the client keeps `alluxio.user.file.metadata.sync.interval` to the configured value
3. If a path is not accessed, the client set `alluxio.user.file.metadata.sync.interval` to `0`

**Urgency**

Urgent

**Comment by secfree** [comment]
`BloomFilter` has false positives, change it to use `Set` here.

**Comment by yyongycy** [comment]
@tcrain @LuQQiu @yuzhu It is about the metadata sync, so we are improving it so far? This should be addressed with the new metadata sync mechanism, right?

This describes a client side solution, it looks like working, but if it is solved in the server side, this should not be a problem. @tcrain Please correct me if anything is wrong.

I am also wondering whether this test is a normal use case or just a test case?

**Comment by Haoning-Sun** [comment]
I have one idea. If getattr must be after list operation, we can consider not setting `alluxio.user.file.metadata.sync.interval` to `0` for getattr(getStatus). After setting  `alluxio.user.file.metadata.sync.interval` to `0` for listStatus, the listStatus triggered by list directory will perform metadata synchronization. This may be applicable to some scenarios.

**Comment by secfree** [comment]
> I am also wondering whether this test is a normal use case or just a test case?
@yyongycy  this is a real use case.

**Comment by LuQQiu** [comment]
@secfree current Alluxio supports NEVER SYNC, ALWAYS SYNC, RESYNC AFTER INTERVAL, here is requesting SYNC ONCE?
Is the workload requires SYNC ONCE be supported automatically and manually? or only automatically or only manually is fine?

**Comment by yuzhu** [comment]
Have you turned on client side metadata cache? or are we discussing server side metadata sync?

**Comment by secfree** [comment]
> current Alluxio supports NEVER SYNC, ALWAYS SYNC, RESYNC AFTER INTERVAL, here is requesting SYNC ONCE?

Here is requesting SYNC ONCE first and then RESYNC AFTER INTERVAL, this can replace using ALWAYS SYNC in some cases.

> Is the workload requires SYNC ONCE be supported automatically and manually? or only automatically or only manually is fine?

Better to be automatically with a on/off parameter. Manually is not expected, as the whole pipeline is scheduled. Once the upstream finished, the downstream will launch.

@LuQQiu

**Comment by secfree** [comment]
> Have you turned on client side metadata cache? or are we discussing server side metadata sync?

No, I haven't turned on client side metadata cache. 
What I propose here is to use the configuration `alluxio.user.file.metadata.sync.interval` at client side to control metadata sync at server side.
@yuzhu

**Comment by yyongycy** [comment]
@elega yimin has a fix, maybe his fix can hit this issue.

**Comment by elega** [comment]
yeah make sure you cherry pick this PR on top of the current 2.9.x build and turn on alluxio.master.metadata.concurrent.sync.dedup.
 Let me know if this solves your issue or not.

**Comment by secfree** [comment]
> alluxio.master.metadata.concurrent.sync.dedup

This looks interesting. We will try and feedback if any updates. Thanks all.

**Comment by secfree** [comment]
After backporting #17244 , the performance of getFileInfo/listStatus are very good now. So I will just close this issue.
